[SPARK-11140] [core] Transfer files using network lib when using NettyRpcEnv.

vanzin opened the pull request:
This change abstracts the code that serves jars / files to executors so that
each RpcEnv can have its own implementation; the akka version uses the existing
HTTP-based file serving mechanism, while the netty versions uses the new
stream support added to the network lib, which makes file transfers benefit
from the easier security configuration of the network library, and should also
reduce overhead overall.

The change includes a small fix to TransportChannelHandler so that it propagates
user events to downstream handlers.

Commit:
[SPARK-11140] [core] Transfer files using network lib when using NettyRpcEnv.

This change abstracts the code that serves jars / files to executors so that
each RpcEnv can have its own implementation; the akka version uses the existing
HTTP-based file serving mechanism, while the netty versions uses the new
stream support added to the network lib, which makes file transfers benefit
from the easier security configuration of the network library, and should also
reduce overhead overall.

The change includes a small fix to TransportChannelHandler so that it propagates
user events to downstream handlers.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

vanzin commented:
/cc @rxin @zsxwing

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45292 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45292/consoleFull)** for PR 9530 at commit [`3f8209f`](https://github.com/apache/spark/commit/3f8209fe664836242ab849d4e8503c4443798556).

SparkQA commented:
**[Test build #45292 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45292/consoleFull)** for PR 9530 at commit [`3f8209f`](https://github.com/apache/spark/commit/3f8209fe664836242ab849d4e8503c4443798556).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45293/consoleFull)** for PR 9530 at commit [`3f8209f`](https://github.com/apache/spark/commit/3f8209fe664836242ab849d4e8503c4443798556).

SparkQA commented:
**[Test build #45293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45293/consoleFull)** for PR 9530 at commit [`3f8209f`](https://github.com/apache/spark/commit/3f8209fe664836242ab849d4e8503c4443798556).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Merge branch 'master' into SPARK-11140

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45383/consoleFull)** for PR 9530 at commit [`87b7a91`](https://github.com/apache/spark/commit/87b7a9174be76646ef814b42dc7d0172ccc23552).

SparkQA commented:
**[Test build #45383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45383/consoleFull)** for PR 9530 at commit [`87b7a91`](https://github.com/apache/spark/commit/87b7a9174be76646ef814b42dc7d0172ccc23552).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class JavaAFTSurvivalRegressionExample `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45402/consoleFull)** for PR 9530 at commit [`87b7a91`](https://github.com/apache/spark/commit/87b7a9174be76646ef814b42dc7d0172ccc23552).

Commit:
Change the file retrieval API.

This reuses more of the existing code in Utils, and also makes some
subsequent changes built on top of this API possible, at the expense
of some efficiency (using input streams instead of channels).

If desired the Utils class can later be changed to use channels to
regain the lost efficiency.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45402/consoleFull)** for PR 9530 at commit [`87b7a91`](https://github.com/apache/spark/commit/87b7a9174be76646ef814b42dc7d0172ccc23552).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing commented on core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala:
```diff
@@ -501,13 +625,15 @@ private[netty] class NettyRpcHandler(
   override def exceptionCaught(cause: Throwable, client: TransportClient): Unit = {
     val addr = client.getChannel.remoteAddress().asInstanceOf[InetSocketAddress]
     if (addr != null) {
-      val clientAddr = RpcAddress(addr.getHostName, addr.getPort)
```
Why reverted it back to the old codes?

vanzin replied:
Sorry, I don't follow. The line was moved inside the new conditions, nothing's really changing.

zsxwing replied:
You added a new if `if (clients.containsKey(client)) {`. There was a similar logic here: https://github.com/apache/spark/pull/9210/files#diff-0c89b4a60c30a7cd2224bb64d93da942L477 but you removed it. Now you are adding it back.

vanzin replied:
It's explained in the following comment:
https://github.com/apache/spark/pull/9530/files#diff-0c89b4a60c30a7cd2224bb64d93da942R85

zsxwing replied:
Got it. Thanks!

zsxwing commented:
LGTM. I noticed that `Utils.fetchFile` cannot handle special chars (e.g., `%`, ` `) in file name. But since the http server has the same issue, it should not block this PR.

vanzin commented on core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala:
```diff
@@ -302,6 +325,105 @@ private[netty] class NettyRpcEnv(
+
+  private class TimeoutHandler(client: TransportClient) extends ChannelInboundHandlerAdapter {
+
+    override def userEventTriggered(ctx: ChannelHandlerContext, evt: Object): Unit = {
```
So, this is a little bit racy, since the timeout might trigger when another thread is preparing to download a file. I'll fix this and update the PR.

Commit:
Handle race in timeout handling.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45446 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45446/consoleFull)** for PR 9530 at commit [`a222c03`](https://github.com/apache/spark/commit/a222c0344738ec1bd19e6cffea83d11d1b0ad2fc).

zsxwing commented on core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala:
```diff
@@ -302,6 +325,138 @@ private[netty] class NettyRpcEnv(
+    source
+  }
+
+  private def fileDownloadClient(host: String, port: Int): TransportClient = synchronized {
```
So if dowloading from different hosts (different tasks in the same Executor), they need to connect one by one. Right?

vanzin replied:
Downloading from different hosts mean different clients. But different tasks will still all download from the same host (the driver).

zsxwing replied:
Oh, right. All tasks will download from the same driver.

zsxwing replied:
Just found another case: `fetchFiles` is also used in `DriverRunner` in the `Worker`.

vanzin replied:
That's fine. What's your worry here?

zsxwing replied:
On second thought I think that's fine since downloading are still in parallel.

SparkQA commented:
**[Test build #45446 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45446/consoleFull)** for PR 9530 at commit [`a222c03`](https://github.com/apache/spark/commit/a222c0344738ec1bd19e6cffea83d11d1b0ad2fc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45446/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-11140

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45540 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45540/consoleFull)** for PR 9530 at commit [`a89e665`](https://github.com/apache/spark/commit/a89e6655e74209d9b89f5dcb2b1b8ed8e3399d42).

zsxwing commented on core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala:
```diff
@@ -302,6 +325,138 @@ private[netty] class NettyRpcEnv(
+          val socketAddr = client.getChannel().remoteAddress().asInstanceOf[InetSocketAddress]
+          val address = RpcAddress(socketAddr.getHostName(), socketAddr.getPort())
+          ctx.close()
+          NettyRpcEnv.this.synchronized {
```
The lock order here is in reverse order of `fileDownloadClient`. A potential dead-lock issue?

vanzin replied:
Good catch, fixed.

Commit:
Fix locking.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45547 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45547/consoleFull)** for PR 9530 at commit [`71ac0cf`](https://github.com/apache/spark/commit/71ac0cfdb25d7bb29e180c704a8eac840c7fccd4).

SparkQA commented:
**[Test build #45540 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45540/consoleFull)** for PR 9530 at commit [`a89e665`](https://github.com/apache/spark/commit/a89e6655e74209d9b89f5dcb2b1b8ed8e3399d42).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45540/
Test PASSed.

Commit:
Doc updates.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45552 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45552/consoleFull)** for PR 9530 at commit [`11f61a8`](https://github.com/apache/spark/commit/11f61a88e4a9a1a1a17572c8579143fc856667d9).

SparkQA commented:
**[Test build #45552 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45552/consoleFull)** for PR 9530 at commit [`11f61a8`](https://github.com/apache/spark/commit/11f61a88e4a9a1a1a17572c8579143fc856667d9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45552/
Test PASSed.

Commit:
Doc consistency fix.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45570/
Test FAILed.

SparkQA commented:
**[Test build #45547 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45547/consoleFull)** for PR 9530 at commit [`71ac0cf`](https://github.com/apache/spark/commit/71ac0cfdb25d7bb29e180c704a8eac840c7fccd4).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45547/
Test FAILed.

Commit:
Merge branch 'master' into SPARK-11140

SparkQA commented:
**[Test build #45904 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45904/consoleFull)** for PR 9530 at commit [`908a051`](https://github.com/apache/spark/commit/908a05121da0a5957845d4451e3841e6faa71493).

SparkQA commented:
**[Test build #45904 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45904/consoleFull)** for PR 9530 at commit [`908a051`](https://github.com/apache/spark/commit/908a05121da0a5957845d4451e3841e6faa71493).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class AppendColumns[T, U](`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45904/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #46041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46041/consoleFull)** for PR 9530 at commit [`908a051`](https://github.com/apache/spark/commit/908a05121da0a5957845d4451e3841e6faa71493).

SparkQA commented:
**[Test build #46041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46041/consoleFull)** for PR 9530 at commit [`908a051`](https://github.com/apache/spark/commit/908a05121da0a5957845d4451e3841e6faa71493).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46041/
Test FAILed.

vanzin commented:
pyspark tests, always flaky... retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46096/
Test FAILed.

zsxwing commented:
Let me take a look at `test_kafka_direct_stream_foreach_get_offsetRanges`

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46098/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #46099 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46099/consoleFull)** for PR 9530 at commit [`908a051`](https://github.com/apache/spark/commit/908a05121da0a5957845d4451e3841e6faa71493).

zsxwing commented:
/cc @rxin is it too risky if adding this one to 1.6?

rxin commented:
Does it buy us anything to add this to 1.6?

vanzin commented:
sasl encryption for file transfers (instead of having to configure ssl for the driver's file server). lower overhead in the driver. but nothing really critical.

rxin commented:
Alright I looked a bit closer at this. Since it is not mission critical and it is touching some mission critical part, I think it's best to not to merge this in 1.6. 

Let's merge this once 1.6 has an rc?

vanzin commented:
What's the problem with merging to master? Isn't that why we have branch-1.6?

I have changes built on top of this that I'm waiting to send for review.

rxin commented:
I'm mostly thinking it'd make it harder if we found bugs that need to fix.

vanzin commented:
Do you mean bugs in this code? Or in 1.6? If in this code, master will eventually go through stabilization also, and at that point if there are bugs, they'll surface. Whether this code goes in now or later doesn't change much in that regard.

rxin commented:
There are still some network related code that's going into 1.6 (one is the config naming thing). I'm concerned they might conflict. Maybe you can just submit more prs basing on previous ones for reviewing?

We should be pretty close to cutting a rc anyway.

vanzin commented:
> Maybe you can just submit more prs basing on previous ones

I don't know of any clean way of doing that in github; they'd end up including all the code in this PR, which is rather ugly / sub-optimal.

I can wait for the config change (I want to push it soon anyway), but I'm not super worried about conflicts. Those are generally not hard to resolve.

Commit:
Merge branch 'master' into SPARK-11140

SparkQA commented:
**[Test build #46261 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46261/consoleFull)** for PR 9530 at commit [`d10fa97`](https://github.com/apache/spark/commit/d10fa97c8458b1ca20adfd8ea594c7cc66e99932).

SparkQA commented:
**[Test build #46261 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46261/consoleFull)** for PR 9530 at commit [`d10fa97`](https://github.com/apache/spark/commit/d10fa97c8458b1ca20adfd8ea594c7cc66e99932).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ALS(override val uid: String) extends Estimator[ALSModel] with ALSParams with Writable `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46261/
Test FAILed.

Commit:
Fix borked merge.

SparkQA commented:
**[Test build #46265 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46265/consoleFull)** for PR 9530 at commit [`7cc83e7`](https://github.com/apache/spark/commit/7cc83e73fc0223c716e73b1ec269c484159092f3).

SparkQA commented:
**[Test build #46265 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46265/consoleFull)** for PR 9530 at commit [`7cc83e7`](https://github.com/apache/spark/commit/7cc83e73fc0223c716e73b1ec269c484159092f3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ALS(override val uid: String) extends Estimator[ALSModel] with ALSParams with Writable `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46265/
Test PASSed.

Commit:
Use separate client factory for file downloads.

This makes the code friendlier to SPARK-11097 (if and when
it's implemented), and also avoids custom timeout handling
code by reusing features of the transport library.

Need to test on a real cluster, though...

SparkQA commented:
**[Test build #46289 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46289/consoleFull)** for PR 9530 at commit [`cfd01bd`](https://github.com/apache/spark/commit/cfd01bddeeedc99b84e756afe97b9731b8e95fc6).

SparkQA commented:
**[Test build #46289 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46289/consoleFull)** for PR 9530 at commit [`cfd01bd`](https://github.com/apache/spark/commit/cfd01bddeeedc99b84e756afe97b9731b8e95fc6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46289/
Test PASSed.

Commit:
Allow separate configs for all network opts in file download client.

But use the rpc config as the default. Also now properly tested on
a real cluster, and verified idle connections are closed.

SparkQA commented:
**[Test build #46349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46349/consoleFull)** for PR 9530 at commit [`0e8e4bb`](https://github.com/apache/spark/commit/0e8e4bb10605d218e8c763ac4fd66bf48113b794).

SparkQA commented:
**[Test build #46349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46349/consoleFull)** for PR 9530 at commit [`0e8e4bb`](https://github.com/apache/spark/commit/0e8e4bb10605d218e8c763ac4fd66bf48113b794).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46349/
Test FAILed.

vanzin commented:
sigh... retest this please.

SparkQA commented:
**[Test build #46361 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46361/consoleFull)** for PR 9530 at commit [`0e8e4bb`](https://github.com/apache/spark/commit/0e8e4bb10605d218e8c763ac4fd66bf48113b794).

SparkQA commented:
**[Test build #46361 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46361/consoleFull)** for PR 9530 at commit [`0e8e4bb`](https://github.com/apache/spark/commit/0e8e4bb10605d218e8c763ac4fd66bf48113b794).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46361/
Test FAILed.

Commit:
Merge branch 'master' into SPARK-11140

SparkQA commented:
**[Test build #46388 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46388/consoleFull)** for PR 9530 at commit [`c5a218d`](https://github.com/apache/spark/commit/c5a218d51c805cb7be8421eb88ed097d1e8dba3b).

SparkQA commented:
**[Test build #46388 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46388/consoleFull)** for PR 9530 at commit [`c5a218d`](https://github.com/apache/spark/commit/c5a218d51c805cb7be8421eb88ed097d1e8dba3b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46388/
Test PASSed.

zsxwing commented:
LGTM. But let's just merge this one to master since it's too risky to add a new feature to 1.6 now.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #46542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46542/consoleFull)** for PR 9530 at commit [`c5a218d`](https://github.com/apache/spark/commit/c5a218d51c805cb7be8421eb88ed097d1e8dba3b).

SparkQA commented:
**[Test build #46542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46542/consoleFull)** for PR 9530 at commit [`c5a218d`](https://github.com/apache/spark/commit/c5a218d51c805cb7be8421eb88ed097d1e8dba3b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46542/
Test PASSed.

vanzin commented:
Merging to master.

JoshRosen commented:
The following two tests have failed for the past three builds in the Hadoop 2.4 Maven Jenkins job:

- org.apache.spark.deploy.yarn.YarnClusterSuite.run Spark in yarn-client mode
- org.apache.spark.deploy.yarn.YarnClusterSuite.user class path first in client mode

According to Jenkins, these failures first started in https://amplab.cs.berkeley.edu/jenkins/job/Spark-Master-Maven-with-YARN/HADOOP_PROFILE=hadoop-2.4,label=spark-test/4223/

Given the set of patches in that changeset, my hunch is that this might have somehow been caused by either #9853 or #9530 (this patch). Branch 1.6 only includes #9853 and seems to be passing these tests, so I therefore believe that this patch may be the culprit (since it's only present in master and not 1.6). @vanzin, could you help me look into this?

squito commented:
Hi @JoshRosen , I was just looking into this.  With some local testing, I could confirm that this patch did lead to those failures, but that it is fixed by https://github.com/apache/spark/pull/9923 (in particular the handling of an empty stream in `TransportResponseHandler`)

squito commented:
(or https://github.com/apache/spark/pull/9941, which is just the bug fixes w/out the other changes ...)
